### PR TITLE
Document for "How to Implement TrainingService in NNI"

### PR DESCRIPTION
**Pull request by SparkSnail**

### CODE REVIEW

File: docs/en_US/HowToImplementTrainingService.md [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    isEarlyStopped?: boolean;
+}
+```
+According to different kinds of implementation, users could put the job detail into a job queue, and keep  fetching the job from the queue and start preparing and running them. Or they could finish preparing and running process in this function, and return job detail after the submit work.
+
+__cancelTrialJob(trialJobId: string, isEarlyStopped?: boolean)__  
+If this function is called, users should execute cancel command to kill a trial job according to trialJobId.
+
+__updateTrialJob(trialJobId: string, form: JobApplicationForm)__  
+This function is called to update the trial job's status, users should execute the corresponding commands to detect the status of a trial process, and change the trial job's status to `RUNNING`, `SUCCEED`, `FAILED` etc.
```
Comment (xuehui1991) [inline docs/en_US/HowToImplementTrainingService.md]: Shall we add a link to the reference part in https://nni.readthedocs.io/en/latest/Reference.html ? To explain the detail of API.

Comment (SparkSnail) [reply docs/en_US/HowToImplementTrainingService.md]: Updated a link to reference.tst

File: docs/en_US/HowToImplementTrainingService.md [hunk 2]
```diff
@@ -0,0 +1,149 @@
+**How to Implement TrainingService in NNI**
+===
+
+## Overview
+TrainingService is a module related to platform management and job schedule in NNI. TrainingService is designed to be easily implemented, we define an abstract class TrainingService as the parent class of all kinds of TrainignService, users just need to  inherit the parent class and complete their own clild chass if they want to implement customized TrainingService.
```
Comment (chicm-ms) [inline docs/en_US/HowToImplementTrainingService.md]: typo: clild chass

File: docs/en_US/HowToImplementTrainingService.md [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+## System architecture
+![](../img/NNIDesign.jpg)
+
+The brief system architecture of NNI is shown in the picture. NNIManager is the core management moudule of system, in charge of calling TrainingService to manage job and the connumication between job and algorithm module. Dispatcher is a message processing center responsible for message dispatch. TrainingService is a module to manage trial jobs, it communicate with nniManager module, and have different instance according to different training platform. For the time being, NNI supports local platfrom, [remote server platfrom](RemoteMachineMode.md), [PAI platfrom](PAIMode.md), [kubeflow platform](KubeflowMode.md) and [FrameworkController platfrom](FrameworkController.md).
```
Comment (chicm-ms) [inline docs/en_US/HowToImplementTrainingService.md]: typo: moudule

Comment (chicm-ms) [inline docs/en_US/HowToImplementTrainingService.md]: typo: connumication

Comment (chicm-ms) [inline docs/en_US/HowToImplementTrainingService.md]: typo: communicate

Comment (chicm-ms) [inline docs/en_US/HowToImplementTrainingService.md]: typo: have

File: docs/en_US/HowToImplementTrainingService.md [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+  | |- sdk
+  | |- webui
+  |- test
+  |- tools
+  | |-nni_annotation
+  | |-nni_cmd
+  | |-nni_gpu_tool
+  | |-nni_trial_tool
+```
+`nni/src/` folder stores the mostly source code of system. The code in this folder is related to NNIManager, TrainingService, SDK, WebUI and other modules. Users could find the abstract class of TrainingService in `nni/src/nni_manager/common/trainingService.ts` file, and they should put their own implemented TrainingService in `nni/src/nni_manager/training_service` folder. If users have implemented their own TrainingService code, they should also supplement the unit test of the code, and place them in `nni/src/nni_manager/training_service/test` folder.
```
Comment (chicm-ms) [inline docs/en_US/HowToImplementTrainingService.md]: typo: mostly

File: docs/en_US/HowToImplementTrainingService.md [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+        this.port = port;
+        this.username = username;
+        this.passwd = passwd;
+        this.sshKeyPath = sshKeyPath;
+        this.passphrase = passphrase;
+        this.gpuReservation = new Map<number, string>();
+    }
+}
+```
+The metadata includes the host address, the username or other configuration related to the platform. Users need to definite their own Metadata format, and set the Matadata instance in this function. This function  is called before the experiment is started, users should set the prerequest configurations and check the healthy of the platform before starting an experiment.
```
Comment (chicm-ms) [inline docs/en_US/HowToImplementTrainingService.md]: typo: definite

Comment (chicm-ms) [inline docs/en_US/HowToImplementTrainingService.md]: typo Matadata

Comment (chicm-ms) [inline docs/en_US/HowToImplementTrainingService.md]: How to follow this instruction?

File: docs/en_US/HowToImplementTrainingService.md (hunk 2 above)
Comment (SparkSnail) [reply docs/en_US/HowToImplementTrainingService.md]: fixed.

File: docs/en_US/HowToImplementTrainingService.md (hunk 3 above)
Comment (SparkSnail) [reply docs/en_US/HowToImplementTrainingService.md]: fixed.

Comment (SparkSnail) [reply docs/en_US/HowToImplementTrainingService.md]: fixed.

Comment (SparkSnail) [reply docs/en_US/HowToImplementTrainingService.md]: fixed

Comment (SparkSnail) [reply docs/en_US/HowToImplementTrainingService.md]: fixed.

File: docs/en_US/HowToImplementTrainingService.md (hunk 4 above)
Comment (SparkSnail) [reply docs/en_US/HowToImplementTrainingService.md]: fixed.

File: docs/en_US/HowToImplementTrainingService.md (hunk 5 above)
Comment (SparkSnail) [reply docs/en_US/HowToImplementTrainingService.md]: fixed to define.

Comment (SparkSnail) [reply docs/en_US/HowToImplementTrainingService.md]: fixed to metadata.

Comment (SparkSnail) [reply docs/en_US/HowToImplementTrainingService.md]: updated to "This function is called before the experiment is started to set the configuration of remote machines."

File: docs/en_US/HowToImplementTrainingService.md [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+    readonly workingDirectory: string;
+    readonly form: JobApplicationForm;
+    readonly sequenceId: number;
+    isEarlyStopped?: boolean;
+}
+```
+According to different kinds of implementation, users could put the job detail into a job queue, and keep  fetching the job from the queue and start preparing and running them. Or they could finish preparing and running process in this function, and return job detail after the submit work.
+
+__cancelTrialJob(trialJobId: string, isEarlyStopped?: boolean)__  
+If this function is called, users should execute cancel command to kill a trial job according to trialJobId.
```
Comment (chicm-ms) [inline docs/en_US/HowToImplementTrainingService.md]: 'execute cancel command'  is too  specific,   can  not be follow.  Suggest to use more general instructions.

File: docs/en_US/HowToImplementTrainingService.md [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+
+__updateTrialJob(trialJobId: string, form: JobApplicationForm)__  
+This function is called to update the trial job's status, users should execute the corresponding commands to detect the status of a trial process, and change the trial job's status to `RUNNING`, `SUCCEED`, `FAILED` etc.
```
Comment (chicm-ms) [inline docs/en_US/HowToImplementTrainingService.md]: "execute the corresponding commands'   maybe too specific,  user  may want to call os APIs which are not commands.

File: docs/en_US/HowToImplementTrainingService.md [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+  | |- sdk
+  | |- webui
+  |- test
+  |- tools
+  | |-nni_annotation
+  | |-nni_cmd
+  | |-nni_gpu_tool
+  | |-nni_trial_tool
+```
+`nni/src/` folder stores the most source code of system. The code in this folder is related to NNIManager, TrainingService, SDK, WebUI and other modules. Users could find the abstract class of TrainingService in `nni/src/nni_manager/common/trainingService.ts` file, and they should put their own implemented TrainingService in `nni/src/nni_manager/training_service` folder. If users have implemented their own TrainingService code, they should also supplement the unit test of the code, and place them in `nni/src/nni_manager/training_service/test` folder.
```
Comment (chicm-ms) [inline docs/en_US/HowToImplementTrainingService.md]: "most source code of system":  suggest to replace "system" with NNI,  we do not say NNI as system.

File: docs/en_US/HowToImplementTrainingService.md (hunk 6 above)
Comment (SparkSnail) [reply docs/en_US/HowToImplementTrainingService.md]: changed to "If this function is called, the trial started by the platform should be canceled. Different kind of platform has diffenent methods to calcel a running job, this function should be implemented according to specific platform."

File: docs/en_US/HowToImplementTrainingService.md (hunk 7 above)
Comment (SparkSnail) [reply docs/en_US/HowToImplementTrainingService.md]: changed to "This function is called to update the trial job's status, trial job's status should be detected according to different platform, and be updated to `RUNNING`, `SUCCEED`, `FAILED` etc."

File: docs/en_US/HowToImplementTrainingService.md (hunk 8 above)
Comment (SparkSnail) [reply docs/en_US/HowToImplementTrainingService.md]: changed to NNI

File: docs/en_US/HowToImplementTrainingService.md [hunk 9]
```diff
@@ -0,0 +1,149 @@
+**How to Implement TrainingService in NNI**
+===
+
+## Overview
+TrainingService is a module related to platform management and job schedule in NNI. TrainingService is designed to be easily implemented, we define an abstract class TrainingService as the parent class of all kinds of TrainignService, users just need to  inherit the parent class and complete their own clild class if they want to implement customized TrainingService.
+
+## System architecture
+![](../img/NNIDesign.jpg)
+
+The brief system architecture of NNI is shown in the picture. NNIManager is the core management module of system, in charge of calling TrainingService to manage job and the communication between job and algorithm module. Dispatcher is a message processing center responsible for message dispatch. TrainingService is a module to manage trial jobs, it communicates with nniManager module, and has different instance according to different training platform. For the time being, NNI supports local platfrom, [remote server platfrom](RemoteMachineMode.md), [PAI platfrom](PAIMode.md), [kubeflow platform](KubeflowMode.md) and [FrameworkController platfrom](FrameworkController.md).
```
Comment (chicm-ms) [inline docs/en_US/HowToImplementTrainingService.md]: "job" is too generic,  should be trial jobs
I can not find "algorithm module" in the picture.

Comment (chicm-ms) [inline docs/en_US/HowToImplementTrainingService.md]: In the picture, it is not "remote server", it is "remote",  text should be consistent with picture.

File: docs/en_US/HowToImplementTrainingService.md [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+This function is called to update the trial job's status, trial job's status should be detected according to different platform, and be updated to `RUNNING`, `SUCCEED`, `FAILED` etc.
+
+__getTrialJob(trialJobId: string)__  
+This function returns a trialJob detail instance according to trialJobId.
+
+__listTrialJobs()__  
+Users should put all of trial job detail information into a list, and return the list.
+
+__addTrialJobMetricListener(listener: (metric: TrialJobMetric) => void)__  
+NNI will hold a EventEmitter to get job metrics, if there is new job metrics detected, the EventEmitter will be triggered. Users should start the EventEmitter in this function.
```
Comment (chicm-ms) [inline docs/en_US/HowToImplementTrainingService.md]: typo: a

File: docs/en_US/HowToImplementTrainingService.md (hunk 9 above)
Comment (SparkSnail) [reply docs/en_US/HowToImplementTrainingService.md]: fixed to remote

Comment (SparkSnail) [reply docs/en_US/HowToImplementTrainingService.md]: fixed to trial job, and chagned to "communication between different modules"

File: docs/en_US/HowToImplementTrainingService.md (hunk 10 above)
Comment (SparkSnail) [reply docs/en_US/HowToImplementTrainingService.md]: fixed to an.

**Pull request merged**